libsolidity/ast/AST.cpp

```diff
@@ -136,28 +137,23 @@ FunctionDefinition const* ContractDefinition::receiveFunction() const
 	return nullptr;
 }
 
-vector<EventDefinition const*> const& ContractDefinition::interfaceEvents() const
+vector<EventDefinition const*> const& ContractDefinition::interfaceEvents(bool _ignoreErrors) const
 {
 	return m_interfaceEvents.init([&]{
-		set<string> eventsSeen;
 		vector<EventDefinition const*> interfaceEvents;
-
 		for (ContractDefinition const* contract: annotation().linearizedBaseContracts)
-			for (EventDefinition const* e: contract->events())
-			{
-				/// NOTE: this requires the "internal" version of an Event,
-				///       though here internal strictly refers to visibility,
-				///       and not to function encoding (jump vs. call)
-				auto const& function = e->functionType(true);
-				solAssert(function, "");
-				string eventSignature = function->externalSignature();
-				if (eventsSeen.count(eventSignature) == 0)
-				{
-					eventsSeen.insert(eventSignature);
-					interfaceEvents.push_back(e);
-				}
-			}
+			interfaceEvents += contract->events();
+		solAssert(annotation().creationCallGraph.set() == annotation().deployedCallGraph.set(), "");
+		if (annotation().creationCallGraph.set())
+		{
+			interfaceEvents += (*annotation().creationCallGraph)->emittedEvents;
+			interfaceEvents += (*annotation().deployedCallGraph)->emittedEvents;
```

Comment:
The problem now is that there can be repetitions here:

```
library L {
	event e();
}
contract C {
	constructor() {
		emit L.e();
	}
	function f() public {
		emit L.e();
	}
}
```

The abi would be

```
[
  {
    "inputs": [],
    "stateMutability": "nonpayable",
    "type": "constructor"
  },
  {
    "anonymous": false,
    "inputs": [],
    "name": "e",
    "type": "event"
  },
  {
    "anonymous": false,
    "inputs": [],
    "name": "e",
    "type": "event"
  },
  {
    "inputs": [],
    "name": "f",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  }
]
```

Similarly, also for the following where we would have 3 repetitions.

```
  contract C {
	  event e();
	  constructor() {
		  emit e();
	  }
	  function f() public {
		  emit e();
	  }
  }
```
We really have to remove the duplicates.